"using" with dynamic crashes the compiler if there is a yield inside it

```C#
    class Program
    {
        static void Main(string[] args)
        {
            foreach(var i in Iter())
            {
                System.Console.WriteLine(i);
            }
        }

        static IEnumerable<Task> Iter()
        {
            dynamic d = 123;

            using (var t = D(d))    
            {
                yield return t;
                t.Wait();
            }
        }

        static Task D(dynamic arg)
        {
             return Task.FromResult(1);
        }
    }
```

 error MSB6006: "csc.exe" exited with code 255.

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs
@@ -121,7 +121,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     @checked: false,
                     constantValueOpt: rewrittenExpression.ConstantValue);
 
-                boundTemp = _factory.StoreToTemp(tempInit, out tempAssignment);
+                boundTemp = _factory.StoreToTemp(tempInit, out tempAssignment, kind: SynthesizedLocalKind.Using);
             }
             else
             {
@@ -183,7 +183,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     @checked: false);
 
                 BoundAssignmentOperator tempAssignment;
-                BoundLocal boundTemp = _factory.StoreToTemp(tempInit, out tempAssignment);
+                BoundLocal boundTemp = _factory.StoreToTemp(tempInit, out tempAssignment, kind: SynthesizedLocalKind.Using);
 
                 BoundStatement tryFinally = RewriteUsingStatementTryFinally(usingSyntax, tryBlock, boundTemp);
 

--- a/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDynamicTests.cs
+++ b/src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDynamicTests.cs
@@ -14614,6 +14614,96 @@ class C
 }");
         }
 
+        [WorkItem(5323, "https://github.com/dotnet/roslyn/issues/5323")]
+        [Fact]
+        public void DynamicUsingWithYield1()
+        {
+            var source =
+@"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        foreach(var i in Iter())
+        {
+            System.Console.WriteLine(i);
+        }
+    }
+
+    static IEnumerable<Task> Iter()
+    {
+        dynamic d = 123;
+
+        using (var t = D(d))    
+        {
+            yield return t;
+            t.Wait();
+        }
+    }
+
+    static Task D(dynamic arg)
+    {
+            return Task.FromResult(1);
+    }
+}";
+            var comp = CreateCompilationWithMscorlib45(source, references: new[] { SystemCoreRef, CSharpRef }, options: TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: @"System.Threading.Tasks.Task`1[System.Int32]");
+
+            comp = CreateCompilationWithMscorlib45(source, references: new[] { SystemCoreRef, CSharpRef }, options: TestOptions.DebugExe);
+
+            CompileAndVerify(comp, expectedOutput: @"System.Threading.Tasks.Task`1[System.Int32]");
+        }
+
+        [WorkItem(5323, "https://github.com/dotnet/roslyn/issues/5323")]
+        [Fact]
+        public void DynamicUsingWithYield2()
+        {
+            var source =
+@"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        foreach(var i in Iter())
+        {
+            System.Console.WriteLine(i);
+        }
+    }
+
+    static IEnumerable<Task> Iter()
+    {
+        dynamic d = 123;
+
+        var t = D(d);
+        using (t)    
+        {
+            yield return t;
+            t.Wait();
+        }
+    }
+
+    static Task D(dynamic arg)
+    {
+            return Task.FromResult(1);
+    }
+}";
+            var comp = CreateCompilationWithMscorlib45(source, references: new[] { SystemCoreRef, CSharpRef }, options: TestOptions.ReleaseExe);
+            
+            CompileAndVerify(comp, expectedOutput: @"System.Threading.Tasks.Task`1[System.Int32]");
+
+            comp = CreateCompilationWithMscorlib45(source, references: new[] { SystemCoreRef, CSharpRef }, options: TestOptions.DebugExe);
+
+            CompileAndVerify(comp, expectedOutput: @"System.Threading.Tasks.Task`1[System.Int32]");
+
+        }
+
         #endregion
 
         #region Using

